src/mock/quick_p2p/tests.rs

```diff
@@ -0,0 +1,542 @@
+    let network = Network::new(None);
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    a.connect_to(b.addr());
+    network.poll();
+
+    a.expect_none();
+    b.expect_none();
+}
+
+#[test]
+fn disconnect_incoming_bootstrap_connection() {
+    let network = Network::new(None);
+
+    let a = Agent::node(&network);
+    let mut b = Agent::bootstrapped_node(&network, a.addr());
+    a.expect_connected_to_node(&b.addr());
+
+    b.disconnect_from(a.addr());
+    network.poll();
+
+    a.expect_connection_failure(&b.addr());
+    b.expect_none();
+}
+
+#[test]
+fn disconnect_outgoing_bootstrap_connection() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let b = Agent::bootstrapped_node(&network, a.addr());
+    a.expect_connected_to_node(&b.addr());
+
+    a.disconnect_from(b.addr());
+    network.poll();
+
+    a.expect_none();
+    b.expect_connection_failure(&a.addr());
+}
+
+#[test]
+fn disconnect_outgoing_connection() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    b.disconnect_from(a.addr());
+    network.poll();
+
+    a.expect_connection_failure(&b.addr());
+    b.expect_none();
+}
+
+#[test]
+fn disconnect_incoming_connection() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    a.disconnect_from(b.addr());
+    network.poll();
+
+    a.expect_none();
+    b.expect_connection_failure(&a.addr());
+}
+
+#[test]
+fn send_to_connected_node() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b.addr(), msg.clone());
+    network.poll();
+
+    b.expect_new_message(&a.addr(), &msg);
+}
+
+#[test]
+fn send_to_disconnecting_node() {
+    let network = Network::new(None);
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b.addr(), msg.clone());
+    b.disconnect_from(a.addr());
+    network.poll();
+
+    a.expect_connection_failure(&b.addr());
+    a.expect_unsent_message(&b.addr(), &msg);
+    b.expect_none();
+}
+
+#[test]
+fn send_to_nonexisting_node() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let b_addr = network.gen_addr();
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b_addr, msg.clone());
+    network.poll();
+
+    // Note: the real quick-p2p will only emit `UnsentUserMessage` when a connection to the peer
```

Comment:
That seem a bit weird, so we send a message, but because there is nothing, we will never know it was not sent?

Reply:
Agree, but it's an issue with the real quic-p2p and I wanted to keep the API identical. Might be worth raising it there (if not already raised).

Reply:
As discussed on slack, a TODO is fine with someone else picking up bring the API fix to the mock that will latter be fixed in quick-p2p lib.